_posts/2016-04-06-intro-to-macros.md

```diff
@@ -0,0 +1,111 @@
+- [Understanding Elixir Macros, Part 5 - Reshaping the AST](http://theerlangelist.com/article/macros_5)
+- [Understanding Elixir Macros, Part 6 - In-place Code Generation](http://theerlangelist.com/article/macros_6)
+
+[**_Book_** Metaprogramming Elixir](https://pragprog.com/book/cmelixir/metaprogramming-elixir) by [@chris_mccord](https://twitter.com/chris_mccord) is next step for deep dive into metaprogramming.
+
+>
+Why another tutorial on already well-documented metaprogramming topic ?
+>
+**To revise and refresh something that we would refer time and again when reviewing Model, Router, View extension DSLs**
+
+Extension DSLs will be using below constructs to get the job done :)
+
+- [Module.register_attribute/3](http://elixir-lang.org/docs/stable/elixir/Module.html#register_attribute/3)
+  - Registers an attribute. By registering an [attribute](http://elixir-lang.org/getting-started/module-attributes.html), a developer is able to customize how Elixir will store and accumulate the attribute values.
+- [Module.put_attribute/3](http://elixir-lang.org/docs/stable/elixir/Module.html#put_attribute/3)
+  - Puts an Erlang attribute to the given module with the given key and value
+- [@before_compile](http://elixir-lang.org/docs/stable/elixir/Module.html)
+  - A hook `__before_compile__/1` that will be invoked before the module is compiled
+  - It allows us to inject code into the module when its definition is complete
+- [\_\_using\_\_ hook](http://elixir-lang.org/docs/stable/elixir/Kernel.html#use/2)
+  - Find the documentation under the `examples for the usage, context and best practices` in link above
+  - `use ModuleName` looks and invokes `__using__` macro defined in `ModuleName` module
+- [bind_quoted](http://elixir-lang.org/docs/stable/elixir/Kernel.SpecialForms.html#quote/2)
+  - Find the documentation under the `bind_quoted option` in the link above
+  - By using `bind_quoted`, we can automatically disable unquoting while still injecting the desired variables into the tree
+- [Code.ensure_loaded?/1](http://elixir-lang.org/docs/stable/elixir/Code.html#ensure_loaded?/1)
+  - Ensures the given module is loaded
+- [apply/3](http://elixir-lang.org/docs/stable/elixir/Kernel.html#apply/2)
+  - Invokes the given `function` from `module` with the array of arguments `args`.
+- [Macro.escape/1](http://elixir-lang.org/docs/stable/elixir/Macro.html#escape/2)
+  - Recursively escapes a value so it can be inserted into a syntax tree
+
+#### <a name="last-section">Metaprogramming pattern as used across extensions</a>
+
+There are three parts needed at minimum to create & use an extension effectively:
+
+- Library Code
+- Service Code
+- Consumer Code
+
+An extension and its use with Nectar can be viewed as Producer / Consumer relationship bound by a communication protocol.
+
+**Extension** which want to add a route, say a list of favorites, is a **Producer (Service Code)**.
```

Comment:
which **wants** to add a route, say a list of favorites, is a Producer (Service Code)